server.js

```diff
@@ -1,20 +1,119 @@
 const express = require("express");
 const cors = require("cors");
-
+const moment = require("moment"); // require
+moment().format("YYYY-MM-DD");
 const app = express();
 
 app.use(express.json());
 app.use(cors());
 
 //Use this array as your (in-memory) data store.
 const bookings = require("./bookings.json");
+const { body, validationResult } = require("express-validator");
 
 app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
 // TODO add your routes and helper functions here
+app.post(
+  "/bookings",
+  [body("email").isEmail().withMessage("Please input a valid email")],
+  function (request, response) {
+    const errors = validationResult(request);
+    if (!errors.isEmpty()) {
+      return response.status(422).json({ errors: errors.array() });
+    }
+    const expectedKeys = [
+      "title",
+      "firstName",
+      "surname",
+      "email",
+      "roomId",
+      "checkInDate",
+      "checkOutDate",
+    ];
+    for (let item in request.body) {
+      if (!request.body[item] || !expectedKeys.includes(item)) {
+        response
+          .status(400)
+          .json("Please ensure your are passing the correct fileds");
+        return;
+      }
```

Comment:
This code does not work - you can't iterate through the keys like this. You need to find the correct method for iterating through the keys of a JSON object. However, an easier way to validate would be to use the express-validator library to specify a minimum length for a field, or use schema validation (more advanced)

Reply:
Hi Karen,

I have tested this again on my live version on heroku, and it was respoding as expected! as for the iteration I've checked those references https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in AND https://www.w3schools.com/js/js_json_objects.asp and I think my code contains the same syntanx for the FOR--IN loop.
If I misunderstood your point can you please let me know. also it will be very helpful if you can share with me the Post request that you made and what response did you get. 

Loads of thanks for your time Karen,

Reply:
here is the live version : https://osman-hotel-server.herokuapp.com/